supernovacontroller/sequential/gpio.py

```diff
@@ -0,0 +1,211 @@
+        """
+        responses = None
+        try:
+            responses = self.controller.sync_submit([
+                lambda transfer_id: self.driver.gpioConfigurePin(transfer_id, pin_number, functionality)
+            ])
+        except Exception as e:
+            raise BackendError(original_exception=e) from e
+
+        response_success = responses[0]["name"] == "CONFIGURE GPIO PIN" and self.__check_if_response_is_successful(responses[0])
+
+        if response_success:
+            self.configured_pins[pin_number] = functionality
+        return (response_success, "Success" if response_success else "Configuration failed, error from the Supernova")
+
+    def digital_write(self, pin_number: GpioPinNumber, logic_level: GpioLogicLevel):
+        """
+        Writes a digital logic level to a GPIO pin.
+
+        Args:
+        pin_number (GpioPinNumber): The GPIO pin number to write to.
+        logic_level (GpioLogicLevel): The logic level to write to the GPIO pin.
+
+        Returns:
+        tuple: A tuple containing two elements:
+            - The first element is a Boolean indicating the success (True) or failure (False) of the write operation.
+            - The second element is a string describing the result of the write operation.
+        """
+        responses = None
+        try:
+            responses = self.controller.sync_submit([
+                lambda transfer_id: self.driver.gpioDigitalWrite(transfer_id, pin_number, logic_level)
+            ])
+        except Exception as e:
+            raise BackendError(original_exception=e) from e
+
+        response_success = responses[0]["name"] == "GPIO DIGITAL WRITE" and self.__check_if_response_is_successful(responses[0])
+        return (response_success, "Success" if response_success else "Digital write failed, error from the Supernova")
+
+    def digital_read(self, pin_number: GpioPinNumber):
+        """
+        Reads the digital logic level from a GPIO pin.
+
+        Args:
+        pin_number (GpioPinNumber): The GPIO pin number to read from.
+
+        Returns:
+        tuple: A tuple containing two elements:
+            - The first element is a Boolean indicating the success (True) or failure (False) of the read operation.
+            - The second element is the logic level read from the GPIO pin if successful, or an error message if failed.
+        """
+        responses = None
+        try:
+            responses = self.controller.sync_submit([
+                lambda transfer_id: self.driver.gpioDigitalRead(transfer_id, pin_number)
+            ])
+        except Exception as e:
+            raise BackendError(original_exception=e) from e
+
+        response_success = responses[0]["name"] == "GPIO DIGITAL READ" and self.__check_if_response_is_successful(responses[0])
+        if response_success:
+            return (True, responses[0]["logic_level"])
+        return (False, "Digital read failed, error from the Supernova")
```

Comment:
In general, it's a good practice to first check for error values and break the flow, and then continue with the "happy path".